modules/tensor_mechanics/src/kernels/StressDivergenceBeam.C

```diff
@@ -0,0 +1,319 @@
+
+  params.set<bool>("use_displaced_mesh") = true;
+  return params;
+}
+
+StressDivergenceBeam::StressDivergenceBeam(const InputParameters & parameters)
+  : Kernel(parameters),
+    _component(getParam<unsigned int>("component")),
+    _ndisp(coupledComponents("displacements")),
+    _disp_var(_ndisp),
+    _nrot(coupledComponents("rotations")),
+    _rot_var(_nrot),
+    _force(&getMaterialPropertyByName<RealVectorValue>("forces")),
+    _moment(&getMaterialPropertyByName<RealVectorValue>("moments")),
+    _K11(getMaterialPropertyByName<RankTwoTensor>("Jacobian_11")),
+    _K22(getMaterialPropertyByName<RankTwoTensor>("Jacobian_22")),
+    _K22_cross(getMaterialPropertyByName<RankTwoTensor>("Jacobian_22_cross")),
+    _K21_cross(getMaterialPropertyByName<RankTwoTensor>("Jacobian_12")),
+    _K21(getMaterialPropertyByName<RankTwoTensor>("Jacobian_21")),
+    _original_length(getMaterialPropertyByName<Real>("original_length")),
+    _total_rotation(&getMaterialPropertyByName<RankTwoTensor>("total_rotation")),
+    _zeta(getMaterialProperty<Real>("zeta")),
+    _alpha(getParam<Real>("alpha")),
+    _force_old(getParam<MaterialPropertyName>("zeta") != "0.0" || std::abs(_alpha) > 0.0
+                   ? &getMaterialPropertyOld<RealVectorValue>("forces")
+                   : nullptr),
+    _moment_old(getParam<MaterialPropertyName>("zeta") != "0.0" || std::abs(_alpha) > 0.0
+                    ? &getMaterialPropertyOld<RealVectorValue>("moments")
+                    : nullptr),
+    _total_rotation_old(getParam<MaterialPropertyName>("zeta") != "0.0" || std::abs(_alpha) > 0.0
+                            ? &getMaterialPropertyOld<RankTwoTensor>("total_rotation")
+                            : nullptr),
+    _force_older(std::abs(_alpha) > 0.0 ? &getMaterialPropertyOlder<RealVectorValue>("forces")
+                                        : nullptr),
+    _moment_older(std::abs(_alpha) > 0.0 ? &getMaterialPropertyOlder<RealVectorValue>("moments")
+                                         : nullptr),
+    _total_rotation_older(std::abs(_alpha) > 0.0
+                              ? &getMaterialPropertyOlder<RankTwoTensor>("total_rotation")
+                              : nullptr)
+{
+  if (_ndisp != _nrot)
+    mooseError(
+        "StressDivergenceBeam: The number of displacement and rotation variables should be same.");
+
+  for (unsigned int i = 0; i < _ndisp; ++i)
+    _disp_var[i] = coupled("displacements", i);
+
+  for (unsigned int i = 0; i < _nrot; ++i)
+    _rot_var[i] = coupled("rotations", i);
+}
+
+void
+StressDivergenceBeam::computeResidual()
+{
+  DenseVector<Number> & re = _assembly.residualBlock(_var.number());
+  mooseAssert(re.size() == 2, "Beam element has and only has two nodes.");
+  _local_re.resize(re.size());
+  _local_re.zero();
+
+  RealVectorValue a(3, 0.0);
+  std::vector<RealVectorValue> global_force_res(_test.size(), a);
+  std::vector<RealVectorValue> global_moment_res(_test.size(), a);
```

Comment:
It's expensive to create these vectors every time this function gets called. Create these as member variables that get re-used instead.